### ISSUE TITLE
Duplicate Claims from StubIdP When Configuring Multiple SAML IdP's

### ISSUE THREAD

**Issue by Bryan-Chaney** [KentorIT/authservices#607]
Using the information from issue _Multiple separate SAML-IdPs in IdSrv3 #325_, I have a test project that has multiple OWIN middleware instances of the AuthServices configured. One of the configurations is the Kentor StubIdP running locally. I'm still getting set up with the other IdP so I can't test that one yet. When running the test app, I see the distinct buttons on the sample login page for each configuration and they will redirect to their respective IdP's so the configuration seems sound.

Everything works as intended when authenticating with the test users in StubIdp, except that I receive duplicates of the claims coming back from it. I discovered that if I configure a 3rd IdP, I will then get 3 copies of the returned claims from the StubIdP, so I seem to get a claim copy for each configured AuthService in my startup even though each is configured for separate IdP's.

Any suggestions for why I'm seeing this interesting behavior? 
No doubt something I need to configure in my test app but haven't.

**Comment by explunit** [comment]
Can you post the code you're using to configure these Idp's?

**Comment by Bryan-Chaney** [comment]
The configuration in the ConfigureAuth method follows. I've dummied up the data for the second IdP. 
I'm still learning about the SAML attributeConsumingService configuration so I've changed nothing there from the sample project. 

<pre>
public void ConfigureAuth(IAppBuilder app)
{
    // Configure the db context, user manager and signin manager to use a single instance per request
    app.CreatePerOwinContext(ApplicationDbContext.Create);
    app.CreatePerOwinContext<ApplicationUserManager>(ApplicationUserManager.Create);
    app.CreatePerOwinContext<ApplicationSignInManager>(ApplicationSignInManager.Create);

    // Enable the application to use a cookie to store information for the signed in user
    // and to use a cookie to temporarily store information about a user logging in with a third party login provider
    // Configure the sign in cookie
    app.UseCookieAuthentication(new CookieAuthenticationOptions
	{
		AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie,
		LoginPath = new PathString("/Account/Login"),
		Provider = new CookieAuthenticationProvider
		{
			// Enables the application to validate the security stamp when the user logs in.
			// This is a security feature which is used when you change a password or add an external login to your account.  
			OnValidateIdentity = SecurityStampValidator.OnValidateIdentity<ApplicationUserManager, ApplicationUser>(
			validateInterval: TimeSpan.FromMinutes(30),
			regenerateIdentity: (manager, user) => user.GenerateUserIdentityAsync(manager))
		}
     });
            
	app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie);

	// Load configuration for the Stub IdP
	app.UseKentorAuthServicesAuthentication(CreateAuthServicesOptions("Stub"));

	// Load configuration for other IdP
	app.UseKentorAuthServicesAuthentication(CreateAuthServicesOptions("Other"));

	// Use a common configuration routine
	private KentorAuthServicesAuthenticationOptions CreateAuthServicesOptions(string idpToUse)
    {
        var spOptions = CreateSPOptions(idpToUse);
        var authServicesOptions = new KentorAuthServicesAuthenticationOptions(false)
        {
			SPOptions = spOptions
        };

        // Handle the IDP specific configuration
        EntityId entityId = null;
        string ssoServiceUrl = null;
        string x509VirtualPath = null;

        switch (idpToUse)
        {
			case "Other":
				entityId = new EntityId("https://otheridp.com");
                ssoServiceUrl = "https://otheridp.com/samlserviceurl";
                x509VirtualPath = "~/App_Data/Other.Idp.cer";
                authServicesOptions.AuthenticationType = "Other";
                authServicesOptions.Caption = "SAML via Other IdP";
                break;
            default:
                entityId = new EntityId("http://stubidp.kentor.se/Metadata");
                ssoServiceUrl = "http://stubidp.kentor.se";
                x509VirtualPath = "~/App_Data/Kentor.AuthServices.StubIdp.cer";
                authServicesOptions.AuthenticationType = "KentorStub";
                authServicesOptions.Caption = "SAML via Kentor stubIdP";
				break;
		}

        var idp = new IdentityProvider(entityId, spOptions)
        {
            AllowUnsolicitedAuthnResponse = true,
            Binding = Saml2BindingType.HttpRedirect,
            SingleSignOnServiceUrl = new Uri(ssoServiceUrl)
        };

        idp.SigningKeys.AddConfiguredKey(
			new X509Certificate2(
                HostingEnvironment.MapPath(x509VirtualPath)
            ));

            authServicesOptions.IdentityProviders.Add(idp);

            return authServicesOptions;
	}

	// SP Configuration
	private static SPOptions CreateSPOptions(string idpToUse)
    {
        // Handle the IDP specific configuration
        EntityId entityId = null;
        string modulePath = null;
        string acsName = null;

            switch (idpToUse)
            {
                case "Other":
                    entityId = new EntityId("http://localhost:60159/OtherAuthServices");
                    modulePath = "/SAML/Other";
                    acsName = "OtherAuthServices";
                    break;
                default:
                    entityId = new EntityId("http://localhost:60159/AuthServices");
                    modulePath = "/SAML/KentorStubIdP";
                    acsName = "AuthServices";
                    break;
            }

            // Continue on as normal
            var english = CultureInfo.GetCultureInfo("en-us");

            var organization = new Organization();
            organization.Names.Add(new LocalizedName("CDSI", english));
            organization.DisplayNames.Add(new LocalizedName("Custom Data Solutions, Inc.", english));
            organization.Urls.Add(new LocalizedUri(new Uri("http://www.custdata.com"), english));

            var spOptions = new SPOptions
            {
                EntityId = entityId,
                ModulePath = modulePath,
                ReturnUrl = new Uri("http://localhost:60159/Account/ExternalLoginCallback"),
                Organization = organization
            };

            var attributeConsumingService = new AttributeConsumingService(acsName)
            {
                IsDefault = true,
            };

            attributeConsumingService.RequestedAttributes.Add(
                new RequestedAttribute("urn:someName")
                {
                    FriendlyName = "Some Name",
                    IsRequired = true,
                    NameFormat = RequestedAttribute.AttributeNameFormatUri
                });

            attributeConsumingService.RequestedAttributes.Add(
                new RequestedAttribute("Minimal"));

            spOptions.AttributeConsumingServices.Add(attributeConsumingService);

            spOptions.ServiceCertificates.Add(new X509Certificate2(
                AppDomain.CurrentDomain.SetupInformation.ApplicationBase + "/App_Data/Kentor.AuthServices.Tests.pfx"));

            return spOptions;
    }
}
</pre>

**Comment by AndersAbel** [comment]
This looks really strange. Can you hook the AcsCommandResultCreated notification and set a breakpoint there to see if that gives any clues?

**Comment by Bryan-Chaney** [comment]
As you suggested, I investigated what was going on around the notification and that looks fine. The SAML response is parsed as expected, although it looks like the <saml2:AttributeStatement> node is not evaluated by the parser (or I need to configure something to get it to do so). The ClaimsPrincipal claims are processed out of the response as expected.

In stepping through this activity, I see that I need to be more accurate about describing the behavior I'm seeing. The Local Authority claims are showing up on the claims display page only once. It's the claims processed from the SAML assertions that are duplicating per authentication configuration.

I'm now working on understanding the middleware in the request/response lifecycle but I see that both middleware configurations are processed via the KentorAuthServicesAuthenticationHandler in the response back from the Stub IdP. The AuthenticationHandlers are invoked multiple times post IdP response, even when only 1 IdP is configured. I don't yet understand this behavior but it is not a cause for the claim duplication on the home page.

**Comment by Bryan-Chaney** [comment]
We can close / delete this issue. I think that the actual problem is that I don't understand SAML & parts of the library well enough to use them properly. Thanks for your time.
